packages/expo/cli/start/server/BundlerDevServer.ts

```diff
@@ -0,0 +1,307 @@
+      // Close the server.
+      if (this.instance?.server) {
+        this.instance.server.close((error) => {
+          this.instance = null;
+          if (error) {
+            reject(error);
+          } else {
+            resolve();
+          }
+        });
+      } else {
+        resolve();
+      }
+    });
+  }
+
+  private getUrlCreator() {
+    assert(this.urlCreator, 'Dev server is not running.');
+    return this.urlCreator;
+  }
+
+  public getNativeRuntimeUrl(opts: Partial<CreateURLOptions> = {}) {
+    return this.isDevClient
+      ? this.getUrlCreator().constructDevClientUrl(opts) ?? this.getDevServerUrl()
+      : this.getUrlCreator().constructUrl({ ...opts, scheme: 'exp' });
+  }
+
+  /** Get the URL for the running instance of the dev server. */
+  public getDevServerUrl(options: { hostType?: 'localhost' } = {}): string | null {
+    if (!this.getInstance()?.location) {
+      return null;
+    }
+    const { location } = this.getInstance();
+    if (options.hostType === 'localhost') {
+      return `${location.protocol}://localhost:${location.port}`;
+    }
+    return location?.url ?? null;
+  }
+
+  /** Get the tunnel URL from ngrok. */
+  public getTunnelUrl(): string | null {
+    return this.ngrok?.getActiveUrl?.() ?? null;
+  }
+
+  /** Open the dev server in a runtime. */
+  public async openPlatformAsync(
+    launchTarget: keyof typeof PLATFORM_MANAGERS | 'desktop',
+    resolver: BaseResolveDeviceProps<any> = {}
+  ) {
+    if (launchTarget === 'desktop') {
+      const url = this.getDevServerUrl({ hostType: 'localhost' });
+      await openBrowserAsync(url);
+      return { url };
+    }
+
+    const runtime = this.isTargetingNative() ? (this.isDevClient ? 'custom' : 'expo') : 'web';
+    const manager = this.getPlatformManager(launchTarget);
+    return manager.openAsync({ runtime }, resolver);
+  }
+
+  /** Should use the interstitial page for selecting which runtime to use. */
+  protected shouldUseInterstitialPage(): boolean {
+    return (
+      env.EXPO_ENABLE_INTERSTITIAL_PAGE &&
+      // TODO: >:0
+      // Checks if dev client is installed.
+      !!resolveFrom.silent(this.projectRoot, 'expo-dev-launcher/package.json')
+    );
+  }
+
+  /** Get the URL for opening in Expo Go. */
+  protected getExpoGoUrl(platform: keyof typeof PLATFORM_MANAGERS) {
+    if (this.shouldUseInterstitialPage()) {
+      const loadingUrl =
+        platform === 'emulator'
+          ? this.urlCreator.constructLoadingUrl({}, 'android')
+          : this.urlCreator.constructLoadingUrl({ hostType: 'localhost' }, 'ios');
+      return loadingUrl;
+    }
+
+    return this.urlCreator.constructUrl({ scheme: 'exp' });
+  }
+
+  protected getPlatformManager(platform: keyof typeof PLATFORM_MANAGERS) {
+    if (!this.platformManagers[platform]) {
+      const Manager = PLATFORM_MANAGERS[platform];
+      this.platformManagers[platform] = new Manager(
+        this.projectRoot,
+        this.getInstance()?.location?.port,
```

Comment:
`location` is always non-null if the server instance is defined
```suggestion
        this.getInstance()?.location.port,
```